client/transaction-pool/graph/src/listener.rs

```diff
@@ -101,8 +101,26 @@ impl<H: hash::Hash + traits::Member + Serialize, H2: Clone + fmt::Debug> Listene
 	}
 
 	/// Transaction was pruned from the pool.
-	pub fn pruned(&mut self, header_hash: H2, tx: &H) {
-		debug!(target: "txpool", "[{:?}] Pruned at {:?}", tx, header_hash);
-		self.fire(tx, |watcher| watcher.in_block(header_hash))
+	pub fn pruned(&mut self, block_hash: BlockHash<C>, tx: &H) {
+		debug!(target: "txpool", "[{:?}] Pruned at {:?}", tx, block_hash);
+		self.fire(tx, |s| s.in_block(block_hash.clone()));
+		self.finality_watchers.entry(block_hash).or_default().push(tx.clone());
```

Comment:
I'm afraid that we may end up with `finality_watchers` growing forever, I'd really like to prevent that and only store a fixed amount of blocks here (say 512 for now).
I think we should introduce another status update that would be send to watchers in such case (`FinalityTimeout`).
Roughly:
```
while self.finality_watchers.len() > 512 {
  if let Some((hash, txs)) = self.finality_watchers.pop_front() {
    for tx in txs {
      self.fire(tx, |s| s.finality_timeout(hash));
    }
  }
}
```

Also `finality_watchers` should switch to be a `LinkedHashMap` and `timeout` event should be causing the watcher to be dropped.